### PR TITLE
[22.05] Fix bulk operations visibility

### LINKED ISSUE #14107: [release testing] Undelete und purge bulk operation option shown in the wrong places
Testing #13555. A couple of comments:

- If I delete some datasets, the option to purge is no longer visible. If I want to purge, I have to first undelete, before the purge option reappears. I think it should be possible to purge in a single step from the deleted state.
- Selecting purged datasets still provides the `Undelete` option. This should be removed or greyed out IMO.

### PR DESCRIPTION
**Pull request by davelopez**
Fixes #14107

- The `Permanently delete` option now appears always.
- The `Undelete` option now is hidden when using manual selection (in query selection we cannot know if there would be non-purged datasets selected) and all items are purged.

**Note**: Trying to purge or undelete a purged dataset will now raise an error in the API. This will probably trigger more error dialogs in the client due to *partial success* but it will make it easier to detect when the history will actually change after an operation so we can stop the `Processing...` message correctly.

fix_bulk_ops_visibility

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Follow instructions on #14107

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/services/history_contents.py [hunk 1]
```diff
@@ -1602,11 +1602,13 @@ def _delete(self, item: HistoryItemModel):
...
     def _undelete(self, item: HistoryItemModel):
         if getattr(item, "purged", False):
-            return
+            raise exceptions.ItemDeletionException("This item has been permanently deleted and cannot be recovered.")
         manager = self._get_item_manager(item)
         manager.undelete(item, flush=self.flush)
 
     def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        if getattr(item, "purged", False):
+            raise exceptions.ItemDeletionException("This item was already permanently deleted.")
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/services/history_contents.py]: Maybe this can just be a no-op?
```suggestion
            return
```

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: This is intentional, although it might be a better solution, see the *note* in the PR description. Your suggestion was the initial approach, but there is a corner case...
If we select only purged datasets and then try to purge them in bulk (for whatever reason) if we no-op, the API call will return success for all of them in the response, and then the UI will expect a history update with the updated items... which will never occur and we get the `Processing operation...` display forever or until page refresh...

So in summary, if we no-op it's hard to know if we should wait for history changes or not, this way since we know there will be no change, we can drop the processing message at the cost of displaying an error which is kind of true :sweat_smile:

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: Sounds like we need to track the progress of a batch operation, it seems unfortunate that we have to raise an exception here to work around not having a way to do this currently. Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: > Sounds like we need to track the progress of a batch operation

Yes, that would be the ideal solution.

> Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Sure, that sounds more reasonable, thanks for the suggestion, will do!

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: By the way, is it ok to raise the error when trying to `undelete` a `purged` dataset or should we do the same, no-op, and just touch the `update_time`?

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: No, I'd raise an exception there.

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/services/history_contents.py]: Done! Thanks for the reviews!

**Review by dannon — APPROVED** [review]
